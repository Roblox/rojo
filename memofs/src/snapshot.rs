use serde::{Serialize, Deserialize};
use std::collections::BTreeMap;
use std::fs;
use std::fs::DirEntry;
use std::path::PathBuf;

/// A slice of a tree of files. Can be loaded into an
/// [`InMemoryFs`](struct.InMemoryFs.html).
#[derive(Debug, Serialize, Deserialize)]
#[non_exhaustive]
pub enum VfsSnapshot {
    File {
        contents: Vec<u8>,
    },

    Dir {
        children: BTreeMap<String, VfsSnapshot>,
    },
}

impl VfsSnapshot {
    pub fn file<C: Into<Vec<u8>>>(contents: C) -> Self {
        Self::File {
            contents: contents.into(),
        }
    }

    pub fn dir<K: Into<String>, I: IntoIterator<Item = (K, VfsSnapshot)>>(children: I) -> Self {
        Self::Dir {
            children: children
                .into_iter()
                .map(|(key, value)| (key.into(), value))
                .collect(),
        }
    }

    pub fn empty_file() -> Self {
        Self::File {
            contents: Vec::new(),
        }
    }

    pub fn empty_dir() -> Self {
        Self::Dir {
            children: BTreeMap::new(),
        }
    }

    pub fn from_fs_path(path: &PathBuf) -> Result<Self, PathBuf> {
        if path.is_file() {
            fs::read_to_string(path).ok()
                .map(|content| Self::file(content))
                .ok_or(path.to_owned())
        } else {
            let entries: Result<Vec<DirEntry>, PathBuf> = fs::read_dir(path)
                .map_err(|_| path.to_owned())?
                .map(|entry| entry.map_err(|_| path.to_owned()))
                .into_iter()
                .collect();

            let vfs_entries: Result<Vec<(String, Self)>, PathBuf> = entries?.iter()
                .map(|entry| {
                    let path = entry.path();

                    path.file_name()
                        .and_then(|file_name| file_name.to_str())
                        .ok_or(path.to_owned())
                        .and_then(|file_name| {
                            Self::from_fs_path(&path)
                                .map(|snapshot| (file_name.to_owned(), snapshot))
                        })
                }).into_iter().collect();

            Ok(Self::dir(vfs_entries?))
        }
    }
}
