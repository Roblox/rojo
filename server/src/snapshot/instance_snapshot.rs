//! Defines the structure of an instance snapshot.

use std::{
    borrow::Cow,
    collections::HashMap,
};

use rbx_dom_weak::{RbxTree, RbxId, RbxValue};

/// A lightweight description of what an instance should look like. Attempts to
/// be somewhat memory efficient by borrowing from its source data, indicated by
/// the lifetime parameter, `'source`.
///
// Possible future improvements:
// - Use refcounted/interned strings
// - Replace use of RbxValue with a sum of RbxValue + borrowed value
#[derive(Debug, Clone, PartialEq)]
pub struct InstanceSnapshot<'source> {
    pub snapshot_id: Option<RbxId>,

    pub name: Cow<'source, str>,
    pub class_name: Cow<'source, str>,
    pub properties: HashMap<String, RbxValue>,
    pub children: Vec<InstanceSnapshot<'source>>,

    // TODO: Snapshot source, like a file or a project node?
}

impl<'source> InstanceSnapshot<'source> {
    pub fn get_owned(&'source self) -> InstanceSnapshot<'static> {
        let children: Vec<InstanceSnapshot<'static>> = self.children.iter()
            .map(InstanceSnapshot::get_owned)
            .collect();

        InstanceSnapshot {
            snapshot_id: None,
            name: Cow::Owned(self.name.clone().into_owned()),
            class_name: Cow::Owned(self.class_name.clone().into_owned()),
            properties: self.properties.clone(),
            children,
        }
    }

    pub fn from_tree(tree: &RbxTree, id: RbxId) -> InstanceSnapshot<'static> {
        let instance = tree.get_instance(id)
            .expect("instance did not exist in tree");

        let children = instance.get_children_ids()
            .iter()
            .cloned()
            .map(|id| InstanceSnapshot::from_tree(tree, id))
            .collect();

        InstanceSnapshot {
            snapshot_id: Some(id),
            name: Cow::Owned(instance.name.clone()),
            class_name: Cow::Owned(instance.class_name.clone()),
            properties: instance.properties.clone(),
            children,
        }
    }
}