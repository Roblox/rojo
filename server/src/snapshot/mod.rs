//! This module defines the instance snapshot subsystem of Rojo.
//!
//! It defines a way to define the instance tree of a project as a pure function
//! of the filesystem by providing a lightweight instance 'snapshot' type, a
//! method to generate minimal patches, and a method that applies those patches.
//!
//! The aim with this approach is to reduce the number of bugs that arise from
//! attempting to manually update instances in response to filesystem updates.
//! Instead of surgically identifying what needs to change, we can do rough
//! "damage-painting", running our relatively fast snapshot function over
//! anything that could have changed and running it through a diffing function
//! to minimize the set of real changes.
//!
//! Building out a snapshot reconciler is mostly overkill for scripts, since
//! their relationships are mostly simple and well-defined. It becomes very
//! important, however, when dealing with large opaque model files and
//! user-defined plugins.

#![allow(dead_code)]

mod patch;
mod patch_apply;
mod patch_compute;
mod instance_snapshot;

pub use instance_snapshot::InstanceSnapshot;
pub use patch_apply::apply_patch_set;
pub use patch_compute::compute_patch_set;
pub use patch::*;