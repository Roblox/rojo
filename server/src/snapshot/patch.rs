//! Defines the data structures used for describing instance patches.

use std::collections::HashMap;

use rbx_dom_weak::{RbxValue, RbxId};

use super::InstanceSnapshot;

/// A set of different kinds of patches that can be applied to an RbxTree.
#[derive(Debug, Clone, PartialEq)]
pub struct PatchSet<'a> {
    pub removed_instances: Vec<RbxId>,
    pub added_instances: Vec<PatchAddInstance<'a>>,
    pub updated_instances: Vec<PatchUpdateInstance>,
}

impl<'a> PatchSet<'a> {
    pub fn new() -> PatchSet<'a> {
        PatchSet {
            removed_instances: Vec::new(),
            added_instances: Vec::new(),
            updated_instances: Vec::new(),
        }
    }
}

/// A patch containing an instance that was added to the tree.
#[derive(Debug, Clone, PartialEq)]
pub struct PatchAddInstance<'a> {
    pub parent_id: RbxId,
    pub instance: InstanceSnapshot<'a>,
}

/// A patch indicating that properties (or the name) of an instance changed.
#[derive(Debug, Clone, PartialEq)]
pub struct PatchUpdateInstance {
    pub id: RbxId,
    pub changed_name: Option<String>,

    /// Contains all changed properties. If a property is assigned to `None`,
    /// then that property has been removed.
    pub changed_properties: HashMap<String, Option<RbxValue>>,
}