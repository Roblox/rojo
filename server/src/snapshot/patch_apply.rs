//! Defines the algorithm for applying generated patches.

use rbx_dom_weak::{RbxTree, RbxId, RbxInstanceProperties};

use super::{
    patch::{PatchSet, PatchUpdateInstance},
    InstanceSnapshot,
};

pub fn apply_patch(
    tree: &mut RbxTree,
    patch_set: &PatchSet,
) {
    for removed_id in &patch_set.removed_instances {
        tree.remove_instance(*removed_id);
    }

    for add_patch in &patch_set.added_instances {
        apply_add_child(tree, add_patch.parent_id, &add_patch.instance);
    }

    for update_patch in &patch_set.updated_instances {
        apply_update_child(tree, update_patch);
    }
}

fn apply_add_child(
    tree: &mut RbxTree,
    parent_id: RbxId,
    snapshot: &InstanceSnapshot,
) {
    let properties = RbxInstanceProperties {
        name: snapshot.name.clone().into_owned(),
        class_name: snapshot.class_name.clone().into_owned(),
        properties: snapshot.properties.clone(),
    };

    let id = tree.insert_instance(properties, parent_id);

    for child_snapshot in &snapshot.children {
        apply_add_child(tree, id, child_snapshot);
    }
}

fn apply_update_child(
    tree: &mut RbxTree,
    patch: &PatchUpdateInstance,
) {
    let instance = tree.get_instance_mut(patch.id)
        .expect("Instance referred to by patch does not exist");

    if let Some(name) = &patch.changed_name {
        instance.name = name.clone();
    }

    if let Some(class_name) = &patch.changed_class_name {
        instance.class_name = class_name.clone();
    }

    for (key, property_entry) in &patch.changed_properties {
        match property_entry {
            Some(value) => {
                instance.properties.insert(key.clone(), value.clone());
            }
            None => {
                instance.properties.remove(key);
            }
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use std::{
        borrow::Cow,
        collections::HashMap,
    };

    use maplit::hashmap;
    use rbx_dom_weak::RbxValue;

    #[test]
    fn add_from_empty() {
        let _ = env_logger::try_init();

        let mut tree = RbxTree::new(RbxInstanceProperties {
            name: "Folder".to_owned(),
            class_name: "Folder".to_owned(),
            properties: HashMap::new(),
        });

        let root_id = tree.get_root_id();

        let snapshot = InstanceSnapshot {
            snapshot_id: None,
            name: Cow::Borrowed("Foo"),
            class_name: Cow::Borrowed("Bar"),
            properties: hashmap! {
                "Baz".to_owned() => RbxValue::Int32 { value: 5 },
            },
            children: Vec::new(),
        };

        apply_add_child(&mut tree, root_id, &snapshot);

        let root_instance = tree.get_instance(root_id).unwrap();
        let child_id = root_instance.get_children_ids()[0];
        let child_instance = tree.get_instance(child_id).unwrap();

        assert_eq!(child_instance.name.as_str(), &snapshot.name);
        assert_eq!(child_instance.class_name.as_str(), &snapshot.class_name);
        assert_eq!(&child_instance.properties, &snapshot.properties);
        assert!(child_instance.get_children_ids().is_empty());
    }

    #[test]
    fn update_existing() {
        let _ = env_logger::try_init();

        let mut tree = RbxTree::new(RbxInstanceProperties {
            name: "OldName".to_owned(),
            class_name: "OldClassName".to_owned(),
            properties: hashmap! {
                "Foo".to_owned() => RbxValue::Int32 { value: 7 },
                "Bar".to_owned() => RbxValue::Int32 { value: 3 },
                "Unchanged".to_owned() => RbxValue::Int32 { value: -5 },
            },
        });

        let root_id = tree.get_root_id();

        let patch = PatchUpdateInstance {
            id: root_id,
            changed_name: Some("Foo".to_owned()),
            changed_class_name: Some("NewClassName".to_owned()),
            changed_properties: hashmap! {
                // The value of Foo has changed
                "Foo".to_owned() => Some(RbxValue::Int32 { value: 8 }),

                // Bar has been deleted
                "Bar".to_owned() => None,

                // Baz has been added
                "Baz".to_owned() => Some(RbxValue::Int32 { value: 10 }),
            },
        };

        apply_update_child(&mut tree, &patch);

        let expected_properties = hashmap! {
            "Foo".to_owned() => RbxValue::Int32 { value: 8 },
            "Baz".to_owned() => RbxValue::Int32 { value: 10 },
            "Unchanged".to_owned() => RbxValue::Int32 { value: -5 },
        };

        let root_instance = tree.get_instance(root_id).unwrap();
        assert_eq!(root_instance.name, "Foo");
        assert_eq!(root_instance.class_name, "NewClassName");
        assert_eq!(root_instance.properties, expected_properties);
    }
}