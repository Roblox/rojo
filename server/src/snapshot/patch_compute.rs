//! Defines the algorithm for computing a roughly-minimal patch set given an
//! existing instance tree and an instance snapshot.

use std::collections::{HashMap, HashSet};

use rbx_dom_weak::{RbxTree, RbxId, RbxInstance};

use super::{
    InstanceSnapshot,
    patch::{PatchSet, PatchAddInstance, PatchUpdateInstance},
};

pub fn compute_patch_set<'a>(
    snapshot: &'a InstanceSnapshot,
    tree: &RbxTree,
    id: RbxId,
) -> PatchSet<'a> {
    let mut patch_set = PatchSet::new();

    compute_patch_set_internal(snapshot, tree, id, &mut patch_set);

    patch_set
}

fn compute_patch_set_internal<'a>(
    snapshot: &'a InstanceSnapshot,
    tree: &RbxTree,
    id: RbxId,
    patch_set: &mut PatchSet<'a>,
){
    let instance = tree.get_instance(id)
        .expect("Instance did not exist in tree");

    compute_property_patches(snapshot, instance, patch_set);
    compute_children_patches(snapshot, tree, id, patch_set);
}

fn compute_property_patches(
    snapshot: &InstanceSnapshot,
    instance: &RbxInstance,
    patch_set: &mut PatchSet,
) {
    let mut visited_properties = HashSet::new();
    let mut changed_properties = HashMap::new();

    let changed_name = if snapshot.name == instance.name {
        None
    } else {
        Some(snapshot.name.clone().into_owned())
    };

    let changed_class_name = if snapshot.class_name == instance.class_name {
        None
    } else {
        Some(snapshot.class_name.clone().into_owned())
    };

    for (name, snapshot_value) in &snapshot.properties {
        visited_properties.insert(name.as_str());

        match instance.properties.get(name) {
            Some(instance_value) => {
                if snapshot_value != instance_value {
                    changed_properties.insert(name.clone(), Some(snapshot_value.clone()));
                }
            }
            None => {
                changed_properties.insert(name.clone(), Some(snapshot_value.clone()));
            }
        }
    }

    for name in instance.properties.keys() {
        if visited_properties.contains(name.as_str()) {
            continue;
        }

        changed_properties.insert(name.clone(), None);
    }

    if changed_properties.is_empty() && changed_name.is_none() {
        return;
    }

    patch_set.updated_instances.push(PatchUpdateInstance {
        id: instance.get_id(),
        changed_name,
        changed_class_name,
        changed_properties,
    });
}

fn compute_children_patches<'a>(
    snapshot: &'a InstanceSnapshot,
    tree: &RbxTree,
    id: RbxId,
    patch_set: &mut PatchSet<'a>,
) {
    let instance = tree.get_instance(id)
        .expect("Instance did not exist in tree");

    let instance_children = instance.get_children_ids();

    let mut paired_instances = vec![false; instance_children.len()];

    for snapshot_child in snapshot.children.iter() {
        let matching_instance = instance_children
            .iter()
            .enumerate()
            .find(|(instance_index, instance_child_id)| {
                if paired_instances[*instance_index] {
                    return false;
                }

                let instance_child = tree.get_instance(**instance_child_id)
                    .expect("Instance did not exist in tree");

                if snapshot_child.name == instance_child.name &&
                    instance_child.class_name == instance_child.class_name
                {
                    paired_instances[*instance_index] = true;
                    return true;
                }

                false
            });

        match matching_instance {
            Some((_, instance_child_id)) => {
                compute_patch_set_internal(snapshot_child, tree, *instance_child_id, patch_set);
            }
            None => {
                patch_set.added_instances.push(PatchAddInstance {
                    parent_id: id,
                    instance: snapshot_child.clone(),
                });
            }
        }
    }

    for (instance_index, instance_child_id) in instance_children.iter().enumerate() {
        if paired_instances[instance_index] {
            continue;
        }

        patch_set.removed_instances.push(*instance_child_id);
    }
}