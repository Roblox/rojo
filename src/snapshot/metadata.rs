use std::{fmt, path::PathBuf, sync::Arc};

use serde::{Deserialize, Serialize};

use crate::{path_serializer, project::ProjectNode};

/// Rojo-specific metadata that can be associated with an instance or a snapshot
/// of an instance.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InstanceMetadata {
    /// Whether instances not present in the source should be ignored when
    /// live-syncing. This is useful when there are instances that Rojo does not
    /// manage.
    pub ignore_unknown_instances: bool,

    /// If a change occurs to this instance, the instigating source is what
    /// should be run through the snapshot functions to regenerate it.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub instigating_source: Option<InstigatingSource>,

    /// The paths that, when changed, could cause the function that generated
    /// this snapshot to generate a different snapshot. Paths should be included
    /// even if they don't exist, since the presence of a file can change the
    /// outcome of a snapshot function.
    ///
    /// For example, a file named foo.lua might have these relevant paths:
    /// - foo.lua
    /// - foo.meta.json (even if this file doesn't exist!)
    ///
    /// A directory named bar/ might have these:
    /// - bar/
    /// - bar/init.meta.json
    /// - bar/init.lua
    /// - bar/init.server.lua
    /// - bar/init.client.lua
    /// - bar/default.project.json
    ///
    /// This path is used to make sure that file changes update all instances
    /// that may need updates.
    // TODO: Change this to be a SmallVec for performance in common cases?
    #[serde(serialize_with = "path_serializer::serialize_vec_absolute")]
    pub relevant_paths: Vec<PathBuf>,

    pub context: InstanceContext,
}

impl InstanceMetadata {
    pub fn new() -> Self {
        Self {
            ignore_unknown_instances: false,
            instigating_source: None,
            relevant_paths: Vec::new(),
            context: InstanceContext::default(),
        }
    }

    pub fn ignore_unknown_instances(self, ignore_unknown_instances: bool) -> Self {
        Self {
            ignore_unknown_instances,
            ..self
        }
    }

    pub fn instigating_source(self, instigating_source: impl Into<InstigatingSource>) -> Self {
        Self {
            instigating_source: Some(instigating_source.into()),
            ..self
        }
    }

    pub fn relevant_paths(self, relevant_paths: Vec<PathBuf>) -> Self {
        Self {
            relevant_paths,
            ..self
        }
    }

    pub fn context(self, context: &InstanceContext) -> Self {
        Self {
            context: context.clone(),
            ..self
        }
    }
}

impl Default for InstanceMetadata {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InstanceContext {
    pub path_ignore_glob: Arc<Vec<IgnoreGlob>>,
}

impl Default for InstanceContext {
    fn default() -> Self {
        InstanceContext {
            path_ignore_glob: Arc::new(Vec::new()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct IgnoreGlob {
    pub pattern: String,
}

#[derive(Clone, PartialEq, Serialize, Deserialize)]
pub enum InstigatingSource {
    Path(#[serde(serialize_with = "path_serializer::serialize_absolute")] PathBuf),
    ProjectNode(String, ProjectNode),
}

impl fmt::Debug for InstigatingSource {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            InstigatingSource::Path(path) => write!(formatter, "Path({})", path.display()),
            InstigatingSource::ProjectNode(name, node) => {
                write!(formatter, "ProjectNode({}: {:?}", name, node)
            }
        }
    }
}

impl From<PathBuf> for InstigatingSource {
    fn from(path: PathBuf) -> Self {
        InstigatingSource::Path(path)
    }
}
