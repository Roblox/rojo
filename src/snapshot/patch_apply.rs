//! Defines the algorithm for applying generated patches.

use std::collections::HashMap;

use rbx_dom_weak::{RbxId, RbxInstanceProperties, RbxValue};

use super::{
    patch::{AppliedPatchSet, AppliedPatchUpdate, PatchSet, PatchUpdate},
    InstancePropertiesWithMeta, InstanceSnapshot, RojoTree,
};

pub fn apply_patch_set(tree: &mut RojoTree, patch_set: PatchSet) -> AppliedPatchSet {
    let mut context = PatchApplyContext::default();

    for removed_id in patch_set.removed_instances {
        apply_remove_instance(&mut context, tree, removed_id);
    }

    for add_patch in patch_set.added_instances {
        apply_add_child(&mut context, tree, add_patch.parent_id, add_patch.instance);
    }

    for update_patch in patch_set.updated_instances {
        apply_update_child(&mut context, tree, update_patch);
    }

    finalize_patch_application(context, tree)
}

#[derive(Default)]
struct PatchApplyContext {
    snapshot_id_to_instance_id: HashMap<RbxId, RbxId>,
    properties_to_apply: HashMap<RbxId, HashMap<String, RbxValue>>,
    applied_patch_set: AppliedPatchSet,
}

/// Finalize this patch application, consuming the context, applying any
/// deferred property updates, and returning the finally applied patch set.
///
/// Ref properties from snapshots refer to eachother via snapshot ID. Some of
/// these properties are transformed when the patch is computed, notably the
/// instances that the patch computing method is able to pair up.
///
/// The remaining Ref properties need to be handled during patch application,
/// where we build up a map of snapshot IDs to instance IDs as they're created,
/// then apply properties all at once at the end.
fn finalize_patch_application(context: PatchApplyContext, tree: &mut RojoTree) -> AppliedPatchSet {
    for (id, mut properties) in context.properties_to_apply {
        let mut instance = tree
            .get_instance_mut(id)
            .expect("Invalid instance ID in deferred property map");

        for property_value in properties.values_mut() {
            if let RbxValue::Ref { value: Some(id) } = property_value {
                if let Some(&instance_id) = context.snapshot_id_to_instance_id.get(id) {
                    *property_value = RbxValue::Ref {
                        value: Some(instance_id),
                    };
                }
            }
        }

        *instance.properties_mut() = properties;
    }

    context.applied_patch_set
}

fn apply_remove_instance(context: &mut PatchApplyContext, tree: &mut RojoTree, removed_id: RbxId) {
    match tree.remove_instance(removed_id) {
        Some(_) => context.applied_patch_set.removed.push(removed_id),
        None => {
            log::warn!(
                "Patch application error: Tried to remove instance {} but it did not exist.",
                removed_id
            );
        }
    }
}

fn apply_add_child(
    context: &mut PatchApplyContext,
    tree: &mut RojoTree,
    parent_id: RbxId,
    snapshot: InstanceSnapshot,
) {
    let properties = InstancePropertiesWithMeta {
        properties: RbxInstanceProperties {
            name: snapshot.name.clone().into_owned(),
            class_name: snapshot.class_name.clone().into_owned(),

            // Property assignment is deferred until after we know about all
            // instances in this patch.
            properties: HashMap::new(),
        },
        metadata: snapshot.metadata.clone(),
    };

    let id = tree.insert_instance(properties, parent_id);

    context.applied_patch_set.added.push(id);

    context
        .properties_to_apply
        .insert(id, snapshot.properties.clone());

    if let Some(snapshot_id) = snapshot.snapshot_id {
        context.snapshot_id_to_instance_id.insert(snapshot_id, id);
    }

    for child_snapshot in snapshot.children {
        apply_add_child(context, tree, id, child_snapshot);
    }
}

fn apply_update_child(context: &mut PatchApplyContext, tree: &mut RojoTree, patch: PatchUpdate) {
    let mut applied_patch = AppliedPatchUpdate::new(patch.id);

    if let Some(metadata) = patch.changed_metadata {
        tree.update_metadata(patch.id, metadata.clone());
        applied_patch.changed_metadata = Some(metadata);
    }

    let mut instance = tree
        .get_instance_mut(patch.id)
        .expect("Instance referred to by patch does not exist");

    if let Some(name) = patch.changed_name {
        *instance.name_mut() = name.clone();
        applied_patch.changed_name = Some(name);
    }

    if let Some(class_name) = patch.changed_class_name {
        *instance.class_name_mut() = class_name.clone();
        applied_patch.changed_class_name = Some(class_name);
    }

    for (key, property_entry) in patch.changed_properties {
        match property_entry {
            // Ref values need to be potentially rewritten from snapshot IDs to
            // instance IDs if they referred to an instance that was created as
            // part of this patch.
            Some(RbxValue::Ref { value: Some(id) }) => {
                let new_id = context
                    .snapshot_id_to_instance_id
                    .get(&id)
                    .copied()
                    .unwrap_or(id);

                instance.properties_mut().insert(
                    key.clone(),
                    RbxValue::Ref {
                        value: Some(new_id),
                    },
                );
            }
            Some(ref value) => {
                instance.properties_mut().insert(key.clone(), value.clone());
            }
            None => {
                instance.properties_mut().remove(&key);
            }
        }

        applied_patch.changed_properties.insert(key, property_entry);
    }

    context.applied_patch_set.updated.push(applied_patch)
}

#[cfg(test)]
mod test {
    use super::*;

    use std::{borrow::Cow, collections::HashMap};

    use maplit::hashmap;
    use rbx_dom_weak::RbxValue;

    use super::super::PatchAdd;

    #[test]
    fn add_from_empty() {
        let _ = env_logger::try_init();

        let mut tree = RojoTree::new(InstancePropertiesWithMeta {
            properties: RbxInstanceProperties {
                name: "Folder".to_owned(),
                class_name: "Folder".to_owned(),
                properties: HashMap::new(),
            },
            metadata: Default::default(),
        });

        let root_id = tree.get_root_id();

        let snapshot = InstanceSnapshot {
            snapshot_id: None,
            metadata: Default::default(),
            name: Cow::Borrowed("Foo"),
            class_name: Cow::Borrowed("Bar"),
            properties: hashmap! {
                "Baz".to_owned() => RbxValue::Int32 { value: 5 },
            },
            children: Vec::new(),
        };

        let patch_set = PatchSet {
            added_instances: vec![PatchAdd {
                parent_id: root_id,
                instance: snapshot.clone(),
            }],
            ..Default::default()
        };

        apply_patch_set(&mut tree, patch_set);

        let root_instance = tree.get_instance(root_id).unwrap();
        let child_id = root_instance.children()[0];
        let child_instance = tree.get_instance(child_id).unwrap();

        assert_eq!(child_instance.name(), &snapshot.name);
        assert_eq!(child_instance.class_name(), &snapshot.class_name);
        assert_eq!(child_instance.properties(), &snapshot.properties);
        assert!(child_instance.children().is_empty());
    }

    #[test]
    fn update_existing() {
        let _ = env_logger::try_init();

        let mut tree = RojoTree::new(InstancePropertiesWithMeta {
            properties: RbxInstanceProperties {
                name: "OldName".to_owned(),
                class_name: "OldClassName".to_owned(),
                properties: hashmap! {
                    "Foo".to_owned() => RbxValue::Int32 { value: 7 },
                    "Bar".to_owned() => RbxValue::Int32 { value: 3 },
                    "Unchanged".to_owned() => RbxValue::Int32 { value: -5 },
                },
            },
            metadata: Default::default(),
        });

        let root_id = tree.get_root_id();

        let patch = PatchUpdate {
            id: root_id,
            changed_name: Some("Foo".to_owned()),
            changed_class_name: Some("NewClassName".to_owned()),
            changed_properties: hashmap! {
                // The value of Foo has changed
                "Foo".to_owned() => Some(RbxValue::Int32 { value: 8 }),

                // Bar has been deleted
                "Bar".to_owned() => None,

                // Baz has been added
                "Baz".to_owned() => Some(RbxValue::Int32 { value: 10 }),
            },
            changed_metadata: None,
        };

        let patch_set = PatchSet {
            updated_instances: vec![patch],
            ..Default::default()
        };

        apply_patch_set(&mut tree, patch_set);

        let expected_properties = hashmap! {
            "Foo".to_owned() => RbxValue::Int32 { value: 8 },
            "Baz".to_owned() => RbxValue::Int32 { value: 10 },
            "Unchanged".to_owned() => RbxValue::Int32 { value: -5 },
        };

        let root_instance = tree.get_instance(root_id).unwrap();
        assert_eq!(root_instance.name(), "Foo");
        assert_eq!(root_instance.class_name(), "NewClassName");
        assert_eq!(root_instance.properties(), &expected_properties);
    }
}
