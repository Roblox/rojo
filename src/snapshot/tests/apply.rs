use rbx_dom_weak::RbxInstanceProperties;

use crate::snapshot::{
    apply_patch_set, InstancePropertiesWithMeta, PatchSet, PatchUpdate, RojoTree,
};

#[test]
fn reify_folder() {
    let mut tree = empty_tree();

    let patch_set = PatchSet {
        updated_instances: vec![PatchUpdate {
            id: tree.get_root_id(),
            changed_name: Some("Hello, world!".to_owned()),
            changed_class_name: Some("Folder".to_owned()),
            changed_properties: Default::default(),
            changed_metadata: None,
        }],
        ..Default::default()
    };

    let _applied_patch_set = apply_patch_set(&mut tree, patch_set);

    // TODO: Make assertions about tree using snapshots
    // TODO: make assertions about applied patch set using snapshots
}

fn empty_tree() -> RojoTree {
    RojoTree::new(InstancePropertiesWithMeta {
        properties: RbxInstanceProperties {
            name: "ROOT".to_owned(),
            class_name: "ROOT".to_owned(),
            properties: Default::default(),
        },
        metadata: Default::default(),
    })
}
