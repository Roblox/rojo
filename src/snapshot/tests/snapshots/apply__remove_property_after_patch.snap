---
source: src/snapshot/tests/apply.rs
expression: tree_value
---
id: id-1
name: ROOT
class_name: ROOT
properties: {}
metadata:
  ignore_unknown_instances: false
  relevant_paths: []
children: []